pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/rest/UserService.java

```diff
@@ -442,4 +443,86 @@ private void checkAssociation(List<User> users) throws ProtocolException{
         }
     }
 
+    /**
+     * Create or update a user record, based on the FENCE user profile, which is in JSON format.
+     *
+     * @param node User profile, as it is received from Gen3 FENCE, in JSON format
+     * @return User The actual entity, as it is persisted (if no errors) in the PSAMA database
+     */
+    public User createUserFromFENCEProfile(JsonNode node) {
+        User actual_user = null;
+        try {
+            User user_to_look_up = new User().setSubject("fence|"+node.get("user_id"));
+
+            if (userRepo == null) {
+                throw new RuntimeException("User repo is null");
+
+            }
+
+            actual_user = userRepo.findOrCreate(user_to_look_up);
+
+            actual_user.setSubject("fence|"+node.get("user_id"));
+            actual_user.setEmail(node.get("email").asText());
+            actual_user.setGeneralMetadata(node.asText());
+
+            // Clear current set of roles every time we
+            actual_user.setRoles(new HashSet<>());
+            userRepo.persist(actual_user);
+        } catch (Exception ex) {
+            logger.error("createUserFromFENCEProfile() ERROR:"+ex.getClass().getName());
+            ex.printStackTrace();
+        }
+        return actual_user;
+    }
+
+    /**
+     * Insert or Update the User object's list of Roles in the database.
+     *
+     * @param u The User object the generated Role will be added to
+     * @param roleName Name of the Role
+     * @param roleDescription Description of the Role
+     * @return boolean Whether the Role was successfully added to the User or not
+     */
```

Comment:
I like the naming! haha upsert!

Reply:
This I've just learned from jason. This is an actual thing: https://www.techbeamers.com/mysql-upsert/